### PR TITLE
Pluggable message router in the  bridge

### PR DESCRIPTION
**Pull request by stefwalter**
This starts work on a pluggable message router in the bridge. Where other processes handle certain channel types. 

Channels get routed to other processes. Some of those processes include:
 * ```cockpit-pcp```
 * ```cockpit-ssh```
 * ```cockpit-bridge --privileged```
 * ```cockpit-xdr``` mythical, but stuff like XDR needs such a C implementation, because it uses FD passing quite often.

These routable processes are declared in packages' ```manifest.json```
 * ```cockpit-ssh``` is declared in the ```dashboard``` ```manifest.json```
 * ```cockpit-pcp``` is declared in a new package ```pcp``` ```manifest.json```
 * ```cockpit-bridge --privileged``` stuff is declared in the ```shell``` ```manifest.json```
They are declared like this:

```
/* pkg/pcp/manifest.json */
"bridges" = [
   {
      "match": { "payload": "metrics1" }
      "spawn": [ "cockpit-pcp" ],
   }
]

/* pkg/shell/manifest.json */
"bridges" = [
    {
        "match": { "superuser": null },
        "spawn": [ "pkexec",  "--disable-internal-agent", "cockpit-bridge", "--privileged" ]
    },
    {
        "match": { "superuser": null },
        "environ": [ "SUDO_PROMPT=\n:::\npassword prompt\n:::\n" ],
        "spawn": [ "sudo", "-n", "cockpit-bridge", "--privileged" ],
    }
]

/* src/base1/manifest.json */
"bridges" = [
    {
         "match": { "superuser": "require" },
         "problem": "access-denied"
    }
]

/* pkg/dashboard/manifest.json */
channels = [
    {
        "match": { "host": "*" },
        "spawn": [ "cockpit-ssh" ]
    }
]
```

 * [x] #5915 
 * [x] #5872 
 * [x] #5871 
 * [x] Tests
 * [x] Documentation

### CODE REVIEW

**Comment by stefwalter** [comment]
@petervo FYI ... and interested in your feedback. After looking at the XDR stuff it became really clear that this needs to be pluggable later, rather than each extension requiring a modification to cockpit-bridge itself.

**Comment by stefwalter** [comment]
Added tests, documentation .. and finished the first two commits.

File: pkg/pcp/manifest.json.in [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+    "bridges": [
+        {
+            "match": { "payload": "metrics1" },
+            "spawn": [ "/usr/libexec/cockpit-pcp" ],
+            "x-spawn": [ "@libexecdir@/cockpit-pcp" ]
```
Comment (petervo) [inline pkg/pcp/manifest.json.in]: I think this might be a typo.

**Comment by stefwalter** [comment]
This failure should be easier to diagnose after #5924 

```
Error: g_hash_table_iter_next: assertion 'ri->version == ri->hash_table->version' failed
```

**Comment by petervo** [comment]
needs rebase

File: src/bridge/cockpitpeer.c [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+      g_debug ("%s: trying to handle channel \"%s\" on peer", self->name, channel);
+      cockpit_transport_freeze (self->transport, channel);
+      cockpit_transport_emit_recv (self->transport, NULL, data);
+    }
+
+  return TRUE;
+}
+
+CockpitTransport *
+cockpit_peer_ensure (CockpitPeer *self)
```
Comment (petervo) [inline src/bridge/cockpitpeer.c]: Should probably comment that there is no transfer here.

File: src/bridge/cockpitpeer.c [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+        {
+          cockpit_transport_close (transport, problem);
+        }
+      else
+        {
+          g_debug ("%s: received init message from peer bridge", self->name);
+          self->inited = TRUE;
+          g_hash_table_iter_init (&iter, self->channels);
+          while (g_hash_table_iter_next (&iter, (gpointer *)&key, NULL))
+            cockpit_transport_thaw (self->transport, key);
```
Comment (petervo) [inline src/bridge/cockpitpeer.c]: Does this risk us modifying the hash table while iterating over it?

File: src/bridge/cockpitpeer.c [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+    {
+      channel = l->data;
+
+      /*
+       * If we have a problem code, that either means that we failed
+       * after the peer bridge came up ... or it didn't come up at
+       * all yet. See above. In these cases we close the channel.
+       */
+      if (problem)
+        {
```
Comment (stefwalter) [inline src/bridge/cockpitpeer.c]: Fixed

File: src/bridge/cockpitpeer.c (hunk 2 above)
Comment (stefwalter) [inline src/bridge/cockpitpeer.c]: Done.

File: src/bridge/cockpitpeer.c (hunk 3 above)
Comment (stefwalter) [inline src/bridge/cockpitpeer.c]: Indeed. Fixed.

File: Makefile.am [hunk 5]
```diff
@@ -114,6 +114,8 @@ SUBST_RULE = \
 	-e 's,[@]prefix[@],$(prefix),g' \
 	-e 's,[@]PACKAGE[@],$(PACKAGE),g' \
 	-e 's,[@]VERSION[@],$(VERSION),g' \
+        -e 's,[@]PKEXEC[@],$(PKEXEC),g' \
+        -e 's,[@]SUDO[@],$(SUDO),g' \
```
Comment (stefwalter) [inline Makefile.am]: Fixed.

File: src/bridge/cockpitpeer.c (hunk 3 above)
Comment (stefwalter) [inline src/bridge/cockpitpeer.c]: Made this a queue as well so we can preserve the order.

**Comment by stefwalter** [comment]
Added a fix for check-authorize unexpected journal message.

**Comment by petervo** [comment]
@stefwalter I reviewed and was ready to merge this. But these changes exposed a bug in #5898 that I had just merged. I added a fix here, if you are ok with then this should be good to go in.

**Pull request closed without merging**